### PR TITLE
support decoding non-scalar tuple elements

### PR DESCRIPTION
**Pull request by fmoor**
fixes https://github.com/edgedb/edgedb-go/issues/54

This change enables decoding tuple elements to default types.
- Object and named tuples decode to map[string]interface{}
- tuples to []interface{}
- scalars to their respective types
- array & set to slice of whatever the element type decodes to

### CODE REVIEW

**Comment by fmoor** [comment]
This is based on https://github.com/edgedb/edgedb-go/pull/51. WIP until https://github.com/edgedb/edgedb-go/pull/51 is merged.

File: internal/codecs/array.go [hunk 1]
```diff
@@ -42,20 +42,31 @@ func popArrayCodec(
...
 type Array struct {
-	id    types.UUID
-	child Codec
-	typ   reflect.Type
-	step  int
+	id         types.UUID
+	child      Codec
+	typ        reflect.Type
+	step       int
+	useReflect bool
```
Comment (1st1) [inline internal/codecs/array.go:49]: Can you add some comments? It's unclear what `step` is or why `useReflect` is necessary

**Review by 1st1 — COMMENTED** [review]
overall lgtm but it's unclear to me what step & useReflect do without deep code reading

File: internal/codecs/array.go (hunk 1 above)
Comment (fmoor) [reply internal/codecs/array.go:54]: added :heavy_check_mark:

File: internal/codecs/array.go [hunk 2]
```diff
@@ -45,17 +45,33 @@ type Array struct {
 	id    types.UUID
 	child Codec
 	typ   reflect.Type
-	step  int
+
+	// step is the element width in bytes for a go array of type `Array.typ`.
+	step int
```
Comment (1st1) [inline internal/codecs/array.go:50]: Why is this needed btw? I don't recall having this field in Python or JS implementations.

Comment (fmoor) [reply internal/codecs/array.go:50]: This is needed because decoding is done by writing data to known memory locations based on the out value's type. When there is a slice (an array under the hood) in the data structure the codec needs to know where each array element starts in memory. Something like:

```go
var head unsafe.Pointer

for i := 0; i < elemCount; i++ {
    decodeNextElementToLocation(head + i*step)
}
```
